src/scene/materials/standard-node-material.js

```diff
@@ -0,0 +1,189 @@
+ * @class
+ * @name pc.StandardNodeMaterial
+ * @augments pc.StandardMaterial
+ * @classdesc StandardNodeMaterial is sub class of the StandardMaterial, and adds shader graph
+ * interop functionality.
+ * @param {pc.StandardMaterial} mat - Optional material which is cloned.
+ * @param {any} chunk - Optional shader graph node chunk to be used.
+ */
+function StandardNodeMaterial(mat, chunk) {
+    StandardMaterial.call(this);
+
+    if (mat) {
+        StandardMaterial.prototype._cloneInternal.call(mat, this);
+    }
+
+    this.shaderGraphChunk = chunk;
+
+    // shader graph parameter and switch overrides
+    this._graphParamOverrides = {};
+    this._graphSwitchOverrides = {};
+}
+
+StandardNodeMaterial.prototype = Object.create(StandardMaterial.prototype);
+StandardNodeMaterial.prototype.constructor = StandardNodeMaterial;
+
+Object.assign(StandardNodeMaterial.prototype, {
+
+    /**
+     * @private
+     * @function
+     * @name pc.StandardNodeMaterial#clone
+     * @description Duplicates a Standard Node material. All properties are duplicated except textures
+     * and the shader graph chunk where only the references are copied.
+     * @returns {pc.StandardNodeMaterial} A cloned Standard Node material.
+     */
+    clone: function () {
+        var clone = new StandardNodeMaterial();
+        StandardMaterial.prototype._cloneInternal.call(this, clone);
+
+        clone.shaderGraphChunk = this._shaderGraphChunk;
+
+        clone._graphParamOverrides = Object.assign({}, this._graphParamOverrides);
+        clone._graphSwitchOverrides = Object.assign({}, this._graphSwitchOverrides);
+
+        return clone;
+    },
+
+    overrideGraphSwitch: function (name, value) {
+        if (this.shaderGraphChunk) {
+            var currentValue = this._graphSwitchOverrides[name];// this.shaderGraphChunk.getSwitchValue(name);
+            if (currentValue !== value) {
+                this._graphSwitchOverrides[name] = value;
+                // this.shaderGraphChunk.setSwitchValue(name, value);
+                this.dirtyShader = true;
+            }
+        }
+    },
+
+    /**
+     * @private
+     * @function
+     * @name pc.StandardNodeMaterial#setParameter
+     * @description Sets a shader parameter on a StandardNodeMaterial material.
+     * @param {string} name - The name of the parameter to set.
+     * @param {number|number[]|pc.Texture} data - The value for the specified parameter.
+     */
+    setParameter: function (name, data) {
+        StandardMaterial.prototype.setParameter.call(this, name, data);
+    },
+
+    setParameters: function (device, names, restoreGraphDefaults) {
+        StandardMaterial.prototype.setParameters.call(this, device, names);
+
+        // apply / restore graph chunk overrides
+        if (this.shaderGraphChunk) {
+            if (restoreGraphDefaults === true) {
+                // restore
+                this.shaderGraphChunk.setParameters(device, Object.keys(this._graphParamOverrides), null);
```

Comment:
Object.keys allocates array, which is not ideal. We should either store overrides in an array directly, or just pass the object as is and iterate over it.

Reply:
Good point - maintaining an array alongside the object makes the most sense.

Reply:
refactored